node/network/gossip-support/src/lib.rs

```diff
@@ -107,26 +140,213 @@ impl GossipSupport {
+				FromOverseer::Signal(OverseerSignal::Conclude) => return self,
+			}
+		}
+	}
+
+	/// 1. Determine if the current session index has changed.
+	/// 2. If it has, determine relevant validators
+	///    and issue a connection request.
+	async fn handle_active_leaves<Context>(
+		&mut self,
+		ctx: &mut Context,
+		leaves: impl Iterator<Item = Hash>,
+	) -> Result<(), util::Error>
+	where
+		Context: SubsystemContext<Message = GossipSupportMessage>,
+		Context: overseer::SubsystemContext<Message = GossipSupportMessage>,
+	{
+		// Use active leaves update as tick to see whether we need to report on connectivity.
+		self.check_connectivity();
+
+		for leaf in leaves {
+			let current_index =
+				util::request_session_index_for_child(leaf, ctx.sender()).await.await??;
+			let since_failure = self.last_failure.map(|i| i.elapsed()).unwrap_or_default();
+			let force_request = since_failure >= BACKOFF_DURATION;
+			let leaf_session = Some((current_index, leaf));
+			let maybe_new_session = match self.last_session_index {
+				Some(i) if current_index <= i => None,
+				_ => leaf_session,
+			};
+
+			let maybe_issue_connection =
+				if force_request { leaf_session } else { maybe_new_session };
+
+			if let Some((session_index, relay_parent)) = maybe_issue_connection {
+				let is_new_session = maybe_new_session.is_some();
+				if is_new_session {
+					tracing::debug!(
+						target: LOG_TARGET,
+						%session_index,
+						"New session detected",
+					);
+				}
+
+				let all_authorities = determine_relevant_authorities(ctx, relay_parent).await?;
+				let our_index = ensure_i_am_an_authority(&self.keystore, &all_authorities).await?;
+				let other_authorities = {
+					let mut authorities = all_authorities.clone();
+					authorities.swap_remove(our_index);
+					authorities
+				};
+
+				self.issue_connection_request(ctx, other_authorities).await?;
+
+				if is_new_session {
+					self.last_session_index = Some(session_index);
+					update_gossip_topology(ctx, our_index, all_authorities, relay_parent).await?;
+				}
 			}
 		}
+
+		Ok(())
+	}
+
+	async fn issue_connection_request<Context>(
+		&mut self,
+		ctx: &mut Context,
+		authorities: Vec<AuthorityDiscoveryId>,
+	) -> Result<(), util::Error>
+	where
+		Context: SubsystemContext<Message = GossipSupportMessage>,
+		Context: overseer::SubsystemContext<Message = GossipSupportMessage>,
+	{
+		let num = authorities.len();
+		let mut validator_addrs = Vec::with_capacity(authorities.len());
+		let mut failures = 0;
+		let mut resolved = HashMap::with_capacity(authorities.len());
+		for authority in authorities {
+			if let Some(addrs) =
+				self.authority_discovery.get_addresses_by_authority_id(authority.clone()).await
+			{
+				validator_addrs.push(addrs.clone());
+				resolved.insert(authority, addrs);
+			} else {
+				failures += 1;
+				tracing::debug!(
+					target: LOG_TARGET,
+					"Couldn't resolve addresses of authority: {:?}",
+					authority
+				);
+			}
+		}
```

Comment:
I'm not a fan of duplicating the logic from network-bridge and having a special low-level message for resolved addresses. Have you considered moving the connectivity check into network-bridge?

Reply:
I have and even started that implementation, but it got ugly/cumbersome because we don't have a central place with all needed information in the network bridge. Also when thinking about implementing proper back pressure, I am expecting quite some refactoring in the network bridge, where I will also take this into account. For now I felt most comfortable with the gossip-support implementation.